### PR TITLE
Scaling dynamics

### PR DESCRIPTION
**Pull request by dallan-keylogic**
## Summary/Motivation:

-Add scaling utility that allows users to quickly scale time `DerivativeVars` and time discretization equations with a single global time scaling factor and the state variable scaling factors.
-Add in an error message for the (presumably) case of mixed spatial/temporal derivatives or higher order time derivatives for PETSc.
-Add an option to pass PETSc a `PETScTrajectory` object from integrating at previous time steps so that the new trajectory is concatenated with it. This helps with MPC, where you'll integrate between times, stop to solve an optimal control problem with the new state variables, then integrate again with updated controls.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by dallan-keylogic** [comment]
@Robbybp @blnicho @eslickj Could yinz take a look at this PR sometime soon? I've got another PR coming along that depends on this one.

File: idaes/core/solvers/petsc.py [hunk 1]
```diff
@@ -531,13 +541,17 @@ def petsc_dae_by_time_element(
...
+        variables_prev = [var[t0] for var in time_vars]
+
     with TemporarySubsystemManager(
         to_deactivate=tdisc,
         to_fix=initial_variables + fix_derivs,
     ):
         # Solver time steps
         deriv_diff_map = _get_derivative_differential_data_map(m, time)
-        tj = None  # trajectory data
+        # tj = None  # trajectory data
```
Comment (eslickj) [inline idaes/core/solvers/petsc.py:554]: Go ahead and delete

**Review by eslickj — APPROVED** [review]

File: idaes/core/solvers/tests/test_petsc.py [hunk 2]
```diff
@@ -717,3 +728,220 @@ def test_rp_example5b():
...
+    assert pytest.approx(y2, rel=1e-3) == pyo.value(m.y[m.t.last(), 2])
+    assert pytest.approx(y3, rel=1e-3) == pyo.value(m.y[m.t.last(), 3])
+    assert pytest.approx(y4, rel=1e-3) == pyo.value(m.y[m.t.last(), 4])
+    assert pytest.approx(y5, rel=1e-3) == pyo.value(m.y[m.t.last(), 5])
+    assert pytest.approx(y6, rel=1e-3) == pyo.value(m.y[m.t.last(), 6])
+
+
+# @pytest.mark.unit
+# @pytest.mark.skipif(not petsc.petsc_available(), reason="PETSc solver not available")
+# def test_petsc_read_trajectory():
```
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/core/solvers/tests/test_petsc.py]: It's a copy of an existing test that I was probably going to modify, before deciding to do something else. It's been removed.

**Review by blnicho — APPROVED** [review]

**Review by Robbybp — CHANGES_REQUESTED** [review]
I think I see some bugs in edge cases. Let me know if I missed something.

File: idaes/core/util/scaling.py [hunk 3]
```diff
@@ -743,6 +760,103 @@ def jacobian_cond(m=None, scaled=True, ord=None, pinv=False, jac=None):
...
+                    cont_eq = getattr(
+                        parent_block, state_var.local_name + "_" + tname + "_cont_eq"
+                    )
+                except AttributeError:
+                    cont_eq = None
+
+                idx_set_list = [idx_set for idx_set in state_var.index_set().subsets()]
+                idx_len = len(idx_set_list)
+                try:
+                    idx_time = idx_set_list.index(time_set)
```
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/core/util/scaling.py]: Thank you for this example. I've added a test to ensure this edge case is handled correctly.

File: idaes/core/util/scaling.py [hunk 4]
```diff
@@ -743,6 +760,103 @@ def jacobian_cond(m=None, scaled=True, ord=None, pinv=False, jac=None):
...
+                        s_state = get_scaling_factor(
+                            state_var[idx], default=1, warning=True
+                        )
+                        set_scaling_factor(var[idx], s_state / time_scaling_factor)
+                    s_derivative = get_scaling_factor(var[idx])
+
+                    if idx_len > 1:
+                        t = idx[idx_time]
+                    else:
+                        t = idx
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/core/util/scaling.py]: This is not designed to work on time-indexed blocks. However, I believe DerivativeVars (and their associated states) and all constraints created by DAE will always be explicitly indexed by time.

**Review by Robbybp — APPROVED** [review]
Looks good assuming tests pass.

**Comment by dallan-keylogic** [comment]
@ksbeattie @lbianchi-lbl @bpaul4 Any idea why the examples tests are failing?

**Comment by bpaul4** [comment]
@dallan-keylogic we recently added `idaes-examples` as a package on the examples-pse side to internally import modules for some examples, and it looks like the idaes-pse is not automatically installing it when getting the examples. We should be able to edit `requirements.txt` to do that, so it's installed for users as well when they update IDAES.

**Comment by bpaul4** [comment]
@lbianchi-lbl I see that `requirements-dev.txt` is installing some editable dependencies [here](https://github.com/IDAES/idaes-pse/blob/c7720427818a0907e83777d7e4610f8bf003d418/requirements-dev.txt#L36), we probably need to update this or add it to CI workflow.

I support manually merging @dallan-keylogic's PR since it is unrelated to the issue and all non-examples tests are passing.
